File: homework1/tom_nguyen/q4_BackspaceStringCompare.cpp
#include <iostream>
#include <string>
using namespace std;

//processes a string by applying backspaces (#) and returns the result
string processBackspaces(const string& s) {
    string result;
    for (char c : s) {
        if (c == '#') {
            if (!result.empty()) {
                result.pop_back();
            }
            continue;
        }
        result += c;
    }
    return result;
}

//compares two strings after applying backspaces in O(n + m)
bool backspaceStringCompare(const string& s1, const string& s2) {
    return processBackspaces(s1) == processBackspaces(s2);
}

int main() {
    cout << boolalpha;
    cout << backspaceStringCompare("abcde", "abcde") << endl; // true
    cout << backspaceStringCompare("Uber Career Prep", "u#Uber Careee#r Prep") << endl; // true
    cout << backspaceStringCompare("abcdef###xyz", "abcw#xyz") << endl; // true
    cout << backspaceStringCompare("abcdef###xyz", "abcdefxyz###") << endl; // false

    return 0;
}

//time spent: 12 minutes

File: junze_ke/homework1/q4_BackspaceStringCompare.py
# Time: 40 Minutes 
# Techique: backwards 2 pointers 
def BackspaceStringCompare(a,b):
    a1 = list(a)
    b1 = list(b)

    l = 0
    r = len(a1) - 1

    l1 = 0
    r1 = len(b1) - 1

    while l < r and l1 < r1:
        if l1 < r1 and b1[r1] == "#":
            r1-=2
        elif l < r and a1[r] == "#":
            r-=2
        else:
            if b1[r1] != a1[r]:
                return False
            else:
                r1-=1
                r-=1
    return True

print(BackspaceStringCompare("abcde","abcde")) #True
print(BackspaceStringCompare("Uber Career Prep","u#Uber Careee#r Prep")) #True
print(BackspaceStringCompare("abcdef###xyz","abcdefxyz###")) #False
